src/nnet3/nnet-simple-component.cc

```diff
@@ -1930,6 +1930,594 @@ void NaturalGradientPerElementScaleComponent::Update(
+ the reversed map as a vector of lists.
+ @param[in] forward_indexes is a vector of int32, each of whose
+            elements is between 0 and input_dim - 1.
+ @param[in] input_dim. See definitions of forward_indexes and
+            backward_indexes.
+ @param[out] backward_indexes is a vector of dimension input_dim
+            of lists, The list at (backward_indexes[i]) is a list
+            of all indexes j such that forward_indexes[j] = i.
+*/
+void Convolutional1dComponent::ReverseIndexes(const std::vector<int32> &forward_indexes,
+                                              int32 input_dim,
+                                              std::vector<std::vector<int32> > *backward_indexes) {
+  int32 i, size = forward_indexes.size();
+  int32 reserve_size = 2 + size / input_dim;
+  backward_indexes->resize(input_dim);
+  std::vector<std::vector<int32> >::iterator iter = backward_indexes->begin(),
+    end = backward_indexes->end();
+  for (; iter != end; ++iter)
+    iter->reserve(reserve_size);
+  for (int32 j = 0; j < forward_indexes.size(); j++) {
+    i = forward_indexes[j];
+    KALDI_ASSERT(i < input_dim);
+    (*backward_indexes)[i].push_back(j);
+  }
+}
+
+/*
+ This function transforms a vector of lists into a list of vectors,
+ padded with -1.
+ @param[in] The input vector of lists. Let in.size() be D, and let
+            the longest list length (i.e. the max of in[i].size()) be L.
+ @param[out] The output list of vectors. The length of the list will
+            be L, each vector-dimension will be D (i.e. out[i].size() == D),
+            and if in[i] == j, then for some k we will have that
+            out[k][j] = i. The output vectors are padded with -1
+            where necessary if not all the input lists have the same side.
+*/
+void Convolutional1dComponent::RearrangeIndexes(const std::vector<std::vector<int32> > &in,
+                                                std::vector<std::vector<int32> > *out) {
+  int32 D = in.size();
+  int32 L = 0;
+  for (int32 i = 0; i < D; i++)
+    if (in[i].size() > L)
+      L = in[i].size();
+  out->resize(L);
+  for (int32 i = 0; i < L; i++)
+    (*out)[i].resize(D, -1);
+  for (int32 i = 0; i < D; i++) {
+    for (int32 j = 0; j < in[i].size(); j++) {
+      (*out)[j][i] = in[i][j];
+    }
+  }
+}
+
+// back propagation function
+void Convolutional1dComponent::Backprop(const std::string &debug_info,
+                                        const ComponentPrecomputedIndexes *indexes,
+                                        const CuMatrixBase<BaseFloat> &in_value,
+                                        const CuMatrixBase<BaseFloat> &out_value,
+                                        const CuMatrixBase<BaseFloat> &out_deriv,
+                                        Component *to_update_in,
+                                        CuMatrixBase<BaseFloat> *in_deriv) const {
+  if (in_deriv)
+    static_cast<CuMatrix<BaseFloat>* >(in_deriv)->Resize(out_deriv.NumRows(), InputDim());
```

Comment:
Ouch!!  You should never static_cast CuMatrixBase to CuMatrix like this!!!  How could this code possibly work if it really pointed to a sub-matrix?  And it will, if it's ever used in a real neural net.    I hope you didn't do this in any previous code.
The code requires that in_deriv already have the correct size.  If you want to zero it you can call SetZero().
But actually, don't call SetZero().  The framework is designed so that if it's more natural to add to the derivative, you just set the flag kBackpropAdds, and it's then the responsibility of the calling code to zero the derivative (if needed).  That way, if the derivative needs to be added to something else, you may be able to avoid redundant copies of matrices.  So add the flag kBackpropAdds, and remove this statement.